tests/Dfe.PlanTech.Web.UnitTests/Background/BackgroundTaskHostedServiceTests.cs

```diff
@@ -0,0 +1,124 @@
+﻿using Dfe.PlanTech.Application.Background;
+using Dfe.PlanTech.Web.Background;
+using Microsoft.Extensions.Logging;
+using NSubstitute;
+
+namespace Dfe.PlanTech.Web.UnitTests.Background;
+
+public class BackgroundTaskHostedServiceTests
+{
+    [Fact]
+    public async Task ExecuteAsync_Runs_One_WorkItem_And_Logs()
+    {
+        // Arrange
+        var logger = Substitute.For<ILogger<BackgroundTaskHostedService>>();
+        var queue = Substitute.For<IBackgroundTaskQueue>();
+
+        var ran = false;
+        var cts = new CancellationTokenSource();
+
+        // Dequeue once, then cancel so the loop exits
+        queue.DequeueAsync(Arg.Any<CancellationToken>())
+             .Returns(Task.FromResult<Func<CancellationToken, Task>>(ct =>
+             {
+                 ran = true;
+                 cts.Cancel(); // stop the loop after this item
+                 return Task.CompletedTask;
+             }));
+
+        var sut = new BackgroundTaskHostedService(logger, queue);
+
+        // Act
+        await sut.StartAsync(cts.Token); // StartAsync calls ExecuteAsync under the hood
+
+        // Assert
+        Assert.True(ran);
+        await queue.Received(1).DequeueAsync(Arg.Any<CancellationToken>());
```

Comment:
Apologies if this is a silly question stemming from me being unfamiliar - is this a race condition?

We seem to: 
- queue a task, 
- start the background task hosted service, 
- immediately assert that it ran, and 
- then de-queue the task?

We would have to be _exceptionally_ unlucky to have such a very short task (just toggling `ran` to `true`) so I don't expect any substantive issues here, but one to flag / explicitly add a comment acknowledging this perhaps?

Reply:
Ah, a misunderstanding on my part - I'm not familiar with NSubstitute and didn't recognise the `Receieved()` syntax, sorry!

I wrongly interpreted it as code being executed as opposed to an assertion/check.

Instead, that line is asserting the call to mocked (substituted?) `queue`'s method `DequeueAsync` has happened exactly once (matching the number of tasks/lambdas passed in to the `Returns`).

---

Additionally, the mocked `DequeueAsync` returns a `Task.FromResult` (i.e. there's no _actual_ background processing/async etc. in this test) which negates the race condition concern.


Thanks for your patience with me on this!